src/Uno.UI/UI/Xaml/DependencyObjectStore.cs

```diff
@@ -1162,13 +1162,70 @@ internal void UpdateResourceBindings(bool isThemeChangedUpdate, ResourceDictiona
 			UpdateChildResourceBindings(isThemeChangedUpdate);
 		}
 
+		private bool _isUpdatingChildResourceBindings;
+
 		private void UpdateChildResourceBindings(bool isThemeChangedUpdate)
 		{
-			foreach (var childStore in _childrenStores.Data)
+			if (_isUpdatingChildResourceBindings)
+			{
+				// Some DPs might be creating reference cycles, so we make sure not to enter an infinite loop.
+				return;
+			}
+			if (isThemeChangedUpdate)
+			{
+				try
+				{
+					_isUpdatingChildResourceBindings = true;
+					foreach (var child in GetChildrenDependencyObjects())
+					{
+						if (!(child is IFrameworkElement) && child is IDependencyObjectStoreProvider storeProvider)
+						{
+							storeProvider.Store.UpdateResourceBindings(isThemeChangedUpdate);
+						}
+					}
+				}
+				finally
+				{
+					_isUpdatingChildResourceBindings = false;
+				}
+			}
+		}
+
+		/// <summary>
+		/// Returns all discoverable child dependency objects.
+		/// </summary>
+		/// <remarks>
+		/// This method is potentially slow and should only be used where performance isn't a concern (eg updating resource bindings
+		/// when the app theme changes).
+		/// </remarks>
+		private IEnumerable<DependencyObject> GetChildrenDependencyObjects()
+		{
+			var propertyValues = _properties.GetAllDetails()
+				.Except(_dataContextPropertyDetails, _templatedParentPropertyDetails)
+				.Select(d => GetValue(d));
+			foreach (var propertyValue in propertyValues)
 			{
-				if (!(childStore.ActualInstance is FrameworkElement)) // FrameworkElements are updated separately by traversing the visual tree
+				if (propertyValue is DependencyObject dependencyObject)
 				{
-					childStore.UpdateResourceBindings(isThemeChangedUpdate);
+					yield return dependencyObject;
+				}
+
+				if (propertyValue is IEnumerable<DependencyObject> dependencyObjectCollection &&
+					(propertyValue is ICollection || propertyValue is DependencyObjectCollectionBase)
```

Comment:
I'm not sure to understand this condition (especially the `ICollection` part), could you please add some doc?